`document.execCommand("copy")` only triggers if there is a selection

nt1m opened the pull request:
#### 03472c5b0900124d973e7e31c74d60864ec9224b
<pre>
`document.execCommand(&quot;copy&quot;)` only triggers if there is a selection
<a href="https://bugs.webkit.org/show_bug.cgi?id=156529">https://bugs.webkit.org/show_bug.cgi?id=156529</a>
<a href="https://rdar.apple.com/27792460">rdar://27792460</a>

Reviewed by Ryosuke Niwa.

Take the focused element as clipboard event target if context is not editable as specification mentions:
<a href="https://www.w3.org/TR/clipboard-apis/#fire-a-clipboard-event">https://www.w3.org/TR/clipboard-apis/#fire-a-clipboard-event</a>

To avoid regressing hire.apple.com when landing bug 284630, we also need to unconditionally allow copying from uneditable elements.

* LayoutTests/editing/execCommand/enabling-and-selection-2-expected.txt:
* LayoutTests/editing/execCommand/enabling-and-selection-2.html:
* LayoutTests/editing/execCommand/enabling-and-selection-expected.txt:
* LayoutTests/editing/execCommand/enabling-and-selection.html:
* LayoutTests/editing/pasteboard/copy-text-from-uneditable-element-expected.txt: Added.
* LayoutTests/editing/pasteboard/copy-text-from-uneditable-element.html: Added.
* LayoutTests/editing/pasteboard/gtk/middle-click-onpaste-expected.txt:
* LayoutTests/editing/pasteboard/gtk/middle-click-onpaste.html:
* LayoutTests/imported/w3c/web-platform-tests/editing/other/exec-command-without-editable-element.tentative-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/editing/other/exec-command-without-editable-element.tentative-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/editing/other/exec-command-without-editable-element.tentative-expected.txt:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::canCopy const):
(WebCore::Editor::findEventTargetFromSelection const):
* Tools/TestWebKitAPI/Tests/WebKitGtk/TestWebViewEditor.cpp:
(testWebViewEditorCutCopyPasteNonEditable):

Canonical link: <a href="https://commits.webkit.org/288559@main">https://commits.webkit.org/288559@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc81c5954721c37e0262323c309793d0b46a008d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83724 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3341 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38025 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88796 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34732 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85809 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3432 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11301 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65126 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22962 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86770 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2534 "Found 1 new test failure: fast/forms/ios/file-upload-panel.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76075 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45415 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2455 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30288 "Found 1 new test failure: fast/editing/recursive-reapply-edit-command-crash.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33781 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73489 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31035 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90174 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10989 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73565 "Passed tests") | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11213 "Failed to checkout and rebase branch from PR 38127") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71900 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/72815 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15754 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2313 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12940 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10941 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16413 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10789 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14264 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12561 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/editing/pasteboard/copy-text-from-uneditable-element.html:
```diff
@@ -0,0 +1,40 @@
+<button id=button>Copy foo</button>
+<input id="result">
+<script>
+jsTestIsAsync = true;
```
Can we add a `description()`?

rniwa commented on LayoutTests/editing/pasteboard/copy-text-from-uneditable-element.html:
```diff
@@ -0,0 +1,40 @@
+<!DOCTYPE html>
+<script src="../../resources/js-test.js"></script>
+<script src="../../resources/ui-helper.js"></script>
+<button id=button>Copy foo</button>
```
We should add another button to paste the content so that we could run this test in a browser.

rniwa commented on LayoutTests/editing/pasteboard/copy-text-from-uneditable-element.html:
```diff
@@ -0,0 +1,40 @@
+        result.focus();
+        document.execCommand("Paste");
+        await shouldBecomeEqual("doneWithPaste", "true");
+        shouldBeEqualToString("result.value", "foo");
```
We should move this to the paste event listener so that the validation happens when we run this test in a browser.

rniwa commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -986,7 +986,14 @@ RefPtr<Element> Editor::findEventTargetFrom(const VisibleSelection& selection) c
+    // https://www.w3.org/TR/clipboard-apis/#fire-a-clipboard-event says:
+    // If the context is editable, then
+    // Set target to be the element that contains the start of the visible selection or cursor in document order, or the body element if there is no visible selection or cursor.
+    if (isEditablePosition(document().selection().selection().start()))
```
I thinks we need to do this whenever selection is a range. Not just when it's editable.

nt1m committed:
`document.execCommand("copy")` only triggers if there is a selection
https://bugs.webkit.org/show_bug.cgi?id=156529
rdar://27792460

Reviewed by Ryosuke Niwa.

Take the focused element as clipboard event target if context is not editable as specification mentions:
https://www.w3.org/TR/clipboard-apis/#fire-a-clipboard-event

To avoid regressing hire.apple.com when landing bug 284630, we also need to unconditionally allow copying from uneditable elements.

* LayoutTests/editing/execCommand/enabling-and-selection-2-expected.txt:
* LayoutTests/editing/execCommand/enabling-and-selection-2.html:
* LayoutTests/editing/execCommand/enabling-and-selection-expected.txt:
* LayoutTests/editing/execCommand/enabling-and-selection.html:
* LayoutTests/editing/pasteboard/copy-text-from-uneditable-element-expected.txt: Added.
* LayoutTests/editing/pasteboard/copy-text-from-uneditable-element.html: Added.
* LayoutTests/editing/pasteboard/gtk/middle-click-onpaste-expected.txt:
* LayoutTests/editing/pasteboard/gtk/middle-click-onpaste.html:
* LayoutTests/imported/w3c/web-platform-tests/editing/other/exec-command-without-editable-element.tentative-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/editing/other/exec-command-without-editable-element.tentative-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/editing/other/exec-command-without-editable-element.tentative-expected.txt:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::canCopy const):
(WebCore::Editor::findEventTargetFromSelection const):
* Tools/TestWebKitAPI/Tests/WebKitGtk/TestWebViewEditor.cpp:
(testWebViewEditorCutCopyPasteNonEditable):

Canonical link: https://commits.webkit.org/288559@main
